esphome/components/modbus_controller/modbus_controller.h

```diff
@@ -445,13 +446,46 @@ class ModbusController : public PollingComponent, public modbus::ModbusDevice {
   uint16_t command_throttle_;
 };
 
-/** convert vector<uint8_t> response payload to float
- * @param value float value to cconvert
+/** Convert vector<uint8_t> response payload to float.
+ * @param data payload with data
  * @param item SensorItem object
- * @return float version of the input
+ * @return float value of data
  */
 inline float payload_to_float(const std::vector<uint8_t> &data, const SensorItem &item) {
-  return payload_to_float(data, item.sensor_value_type, item.offset, item.bitmask);
+  int64_t number = payload_to_number(data, item.sensor_value_type, item.offset, item.bitmask);
+
+  float float_value;
+  if (item.sensor_value_type == SensorValueType::FP32 || item.sensor_value_type == SensorValueType::FP32_R) {
+    union {
+      float float_value;
+      uint32_t raw;
+    } raw_to_float;
+    raw_to_float.raw = number;
+    float_value = raw_to_float.float_value;
```

Comment:
```suggestion
    float_value = bit_cast<float>(static_cast<uint32_t>(raw));
```
(might need to include helpers.h if it doesn't already).